src/identity/manager.rs

```diff
@@ -100,140 +98,332 @@ impl Default for Identity {
+        if cfg.concurrency == 0 {
+            panic!("concurrency cannot be 0, operations would block forever");
+        }
+        let (tx, rx) = mpsc::channel(10);
+        let ret = Arc::new(Self {
+            client,
+            time_conv: cfg.time_conv,
+            concurrency: cfg.concurrency,
+            certs: Default::default(),
+            identities: tx,
+        });
+
+        // Process requests in the background. The task will terminate on its own when the
+        // identities channel is dropped.
+        let r = ret.clone();
+        // TODO: simplify the callback expr?
+        tokio::spawn(async move { r.process_requests(rx).await });
+
+        ret
+    }
+
+    // Manages certificate updates. Since all the work is done in a single task, the code is
+    // lock-free. This is OK as the code is I/O bound so we don't need the extra parallelism.
+    async fn process_requests(&self, mut requests: mpsc::Receiver<(Identity, Priority)>) {
+        use futures::stream::FuturesUnordered;
+        use futures::StreamExt;
+        use priority_queue::PriorityQueue;
+
+        // A set of futures refreshing the certificates. Each future completes with the identity for
+        // which it was invoked and a resulting certificate or error.
+        let mut workers = FuturesUnordered::new();
+        // The set of identities for which there are pending requests. Elements of workers and
+        // processing correspond to each other.
+        let mut processing = HashSet::new();
+        // Identities for which we will need to refresh certificates in the future, ordered by the
+        // priority and time at which the refresh needs to happen.
+        //
+        // Note that while the sorting criteria may seem too simple, it is in fact correct due to
+        // the specifics of values inserted. Only Background priority items are ever inserted into
+        // the future, for all other priorities Instant::now() is used as the scheduled time of the
+        // refresh. In other words, at any point in time, there are no high-priority
+        // (not Background) items scheduled to run in the future.
+        let mut pending: PriorityQueue<Identity, PendingPriority> = PriorityQueue::new();
+
+        'main: loop {
+            let next = pending.peek().map(|(_, PendingPriority(_, ts))| *ts);
+            tokio::select! {
+                res = requests.recv() => match res {
+                    Some((id, pri)) => {
+                        if !processing.contains(&id) {
+                            pending.push_increase(id, PendingPriority(pri, Instant::now()));
                         }
-                    }
-                    fetched_certs.get_duration_until_refresh()
-                }
+                    },
+                    None => break 'main,
+                },
+                Some(res) = workers.next() => match res {
+                    Err(_) => break 'main,
```

Comment:
Workers are refresh(...) calls, which can return only watch::error::SendError, which cannot really happen because the recv end of the channel is kept alive via self. Figured no reason to panic.

Though I just realized that the other break cannot happen because the send end of requests is also referenced via self. So this code leaks the background worker - which AFAIR is not a regression, if you want I can submit as-is and fix it later.